builtin/logical/mongodb/path_roles.go

```diff
@@ -0,0 +1,228 @@
+				Type:        framework.TypeString,
+				Description: "MongoDB roles to assign to the users generated for this role.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.UpdateOperation: b.pathRoleCreate,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleHelpSyn,
+		HelpDescription: pathRoleHelpDesc,
+	}
+}
+
+func (b *backend) Role(s logical.Storage, n string) (*roleStorageEntry, error) {
+	entry, err := s.Get("role/" + n)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result roleStorageEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	role, err := b.Role(req.Storage, data.Get("name").(string))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, nil
+	}
+
+	rolesJsonBytes, err := json.Marshal(role.MongoDBRoles.toStandardRolesArray())
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"db":    role.DB,
+			"roles": string(rolesJsonBytes),
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleList(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	entries, err := req.Storage.List("role/")
+	if err != nil {
+		return nil, err
+	}
+
+	return logical.ListResponse(entries), nil
+}
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	name := data.Get("name").(string)
+	if name == "" {
+		return logical.ErrorResponse("Missing name"), nil
+	}
+
+	roleDB := data.Get("db").(string)
+	if roleDB == "" {
+		return logical.ErrorResponse("db parameter is required"), nil
+	}
+
+	// Example roles JSON:
+	//
+	// [ "readWrite", { "role": "readWrite", "db": "test" } ]
+	//
+	// For storage, we convert such an array into a homogeneous array of role documents like:
+	//
+	// [ { "role": "readWrite" }, { "role": "readWrite", "db": "test" } ]
+	//
+	var roles []mongodbRole
+	rolesJson := []byte(data.Get("roles").(string))
+	if len(rolesJson) > 0 {
+		var rolesArray []interface{}
+		err := json.Unmarshal(rolesJson, &rolesArray)
+		if err != nil {
+			return nil, err
+		}
+		for _, rawRole := range rolesArray {
+			switch role := rawRole.(type) {
+			case string:
+				roles = append(roles, mongodbRole{Role: role})
+			case map[string]interface{}:
+				if db, ok := role["db"].(string); ok {
+					if roleName, ok := role["role"].(string); ok {
+						roles = append(roles, mongodbRole{Role: roleName, DB: db})
```

Comment:
What this won't handle is `{ "role": "readWrite" }`. This isn't necessarily a problem, it just forces the user to always use the non-map version when not specifying a DB. I'm merely pointing this out; it could be an easy future enhancement if needed.